### PR TITLE
Fix broken Validator:isValidMatch

### PR DESCRIPTION
**Pull request by qqux**

### CODE REVIEW

**Review by vstelmakh — COMMENTED** [review]
Thank you for the PR @qqux. Some description would be helpful. :blush:

If I understand your idea correctly you want to change Validator to work a bit differently, to always accept matches which looks like domain. This is not the expected approach as `match by TLD` flag indicating if url match without scheme but with valid TLD should be considered as correct match. Otherwise matches without scheme considered as invalid. 

Merging this change will lead to incorrect url highlights for `match by TLD = flase`.

Please explain if I get it wrong.

**Comment by qqux** [comment]
Hello @vstelmakh, thanks for the response.

The current validation method along with its unit tests is quite weird. I'm using your library in my projects and I encountered some real-life examples where it gives incorrect results when tweaking unrelated options:

Example: `matchEmails = true`

```
   email@example.com -> true
https://example.com/ -> true
         example.com -> true
```

All correct. But set `matchEmails = false` and suddenly you have
```
   email@example.com -> false (as expected)
https://example.com/ -> true
         example.com -> false (wrong!)
```

Regarding the unit tests and matching "edge cases" with `$this->getMatch(null, null, null, null)` etc., the existing behavior is already incorrect. Example:

`$this->getMatch('http', null, null, null)` returns `true`. but it's obviously not a valid URL since it has no host. Note that `Validator` already receives strings that are formed like an URL and merely does some additional validation on them, so that's not the issue, same when we give it `$this->getMatch(null, null, null, null)` (because that would equate to an empty string).

If you want to be strict about it, you can add additional checking to `isValidMatch` method to verify it has all required fields present, but in my opinion it's not necessary.

Regarding `matchByTLD`, yes, you're correct. I inadvertently changed its behavior a bit and will fix that.

**Comment by qqux** [comment]
@vstelmakh ok, if this is the desired behavior, I've reduced my PR to just fixing broken handling of e-mails. Note that the original unit tests all pass despite the fix, which is kind of a bad thing.

**Comment by vstelmakh** [comment]
@qqux I'm not able to reproduce this case:
> All correct. But set `matchEmails = false` and suddenly you have
> 
> ```
>    email@example.com -> false (as expected)
> https://example.com/ -> true
>          example.com -> false (wrong!)
> ```

Here is a dataset I used for `ValidatorTest::testIsValidMatch`:
```php
[
    [true, [], [], false, $this->getMatch(null, 'email', 'example.com', 'com'), false],
    [true, [], [], false, $this->getMatch('http', null, 'example.com', 'com'), true],
    [true, [], [], false, $this->getMatch(null, null, 'example.com', 'com'), true],
]
```

For me it works properly. How did you achieved your results?

**Comment by vstelmakh** [comment]
> Regarding the unit tests and matching "edge cases" with `$this->getMatch(null, null, null, null)` etc., the existing behavior is already incorrect. Example:
> 
> `$this->getMatch('http', null, null, null)` returns `true`. but it's obviously not a valid URL since it has no host. Note that `Validator` already receives strings that are formed like an URL and merely does some additional validation on them, so that's not an issue, same when we give it `$this->getMatch(null, null, null, null)` (because that would equate to an empty string).

Definitely cases mentioned above are not proper matches. As `ValidatorTest` is the unit test for `Validator`, there is no benefit to set complete `UrlMatch` objects. For each test case, only the required data is set. For example `Validator` expect match to be valid if there is scheme defined. Setting up full `UrlMatch` object will just blur the focus from the test case itself. That's why only scheme is set.

Take a look to `ValidatorTest::getMatch` also `fullMatch`, `byteOffset` and `url` properties are always omitted. 

There is `MatcherTest` which takes care of testing how `UrlMatch` objects are created form text input.

File: src/Validator/Validator.php [hunk 1]
```diff
@@ -59,8 +59,8 @@ public function isValidMatch(UrlMatch $match): bool
             return $this->isAllowedScheme($scheme);
         }
 
-        if (!$this->matchEmails && $this->isEmail($match)) {
-            return false;
+        if ($this->isEmail($match)) {
+            return $this->matchEmails;
```
Comment (vstelmakh) [inline src/Validator/Validator.php:63]: Could you please explain what issue does it solve?

For me it looks like it behave the same as previous condition. Apart of the fact that previous condition will work faster in cases when `$this->matchEmails = true` as it will not perform unnecessary `$this->isEmail($match)` call. Which is probably most of the library use cases.

**Comment by qqux** [comment]
> @qqux I'm not able to reproduce this case:

Yes, I was writing from memory and I looked at the wrong parameter.

Set `matchTLD = false, matchEmails = true` and then you get this:

```
         example.com -> false (as expected)
   email@example.com -> false (wrong!)
https://example.com/ -> true
```

> For me it looks like it behave the same as previous condition

@vstelmakh it's not the same. It would've been the same if the method returned `true` at the end.

**Comment by vstelmakh** [comment]
Yes, that's expected behavior, as we can't be sure if this is an email or, for example, a filename `email@example.txt`. And if the TLD check is disabled than it considered as false.

In this case only `mailto:email@example.com` will match.

I know it's probably not very obvious. Maybe it make sense to mention this in the documentation :thinking:

**Comment by qqux** [comment]
> In this case only `mailto:email@example.com` will match.

It still won't because of https://github.com/vstelmakh/url-highlight/blob/7dffee7b2679d5d1ea98d0c3beaf316de3559e30/src/Validator/Validator.php#L58

😉

**Comment by vstelmakh** [comment]
@qqux anyway, thanks for bringing this this topic out. I will think how to make it better.
